### ISSUE TITLE
Add `Html.GetEnumSelectList()` helpers

### ISSUE THREAD

**Issue by NTaylorMullen** [aspnet/Mvc#438]
The first select helper `DropDownList` does not include any special-cases for `enum` types.  Need to support at least the scenarios that `EnumHelpers` and the `EnumDropDownList()` and `EnumDropDownListFor()` helpers enable in legacy MVC.

**Comment by dougbu** [comment]
design is fairly simple: when unable to find a select list for a non-`[Flags]` `enum`, generate a collection of `SelectListItem`s based on the `enum`'s type.  this generation will be analogous to `EnumHelper.GetSelectList(Type)` from MVC 5.  (that class' handling of the current expression result will _not_ be included.  #2027 covers a similar but broader enhancement.)

minor complication for `<select/>` tag helper is that the Razor source may contain `<option/>` elements covering at least some of the `enum` fields.  place a `SelectedInformation` instance into the `FormContext` instead of the current `ICollection<string>` collection of selected values.  this allows contained `<option/>` tag helpers to inform the `<select/>` tag helper of explicit values.  the `<select/>` tag helper will use this data to eliminate duplicates.  also reqruires the `<select/>` tag helper to call `TagHelperContext.GetChildContentAsync()`.

much of the above is shared with the #602 and #2027 fixes.

TL;DR
this enhancement will not support `[Flags]` enums.  `@Html.ListBox[For]()` does not support non-collections.  and, while the form value provider correctly maps multiple values from a `<select/>` element to an `object` array, model binding accepts only the first element of that array.  so, no work to enable part of a scenario.

details of preconditions:
- `DefaultHtmlGenerator.GenerateSelect()` is called with `selectList: null`
- no `IEnumerable<SelectListItem>` is found in `ViewData`
- `ModelMetadata.ModelType.GetTypeInfo().IsEnum` or `Nullable.GetUnderlyingType(ModelMetadata.ModelType).GetTypeInfo().IsEnum`
- `enumType.GetTypeInfo().GetCustomAttribute<FlagsAttribute>() == null`

**Comment by dougbu** [comment]
based on conversations with @Eilon, have simplified the API surface to a helper method that returns a select list for a given `enum` type.  e.g. `public IEnumerable<SelectListItem> GetSelectList<TEnum>() where TEnum : struct`.  so basically redo the `EnumHelper` method from MVC 5.

could include `... GetSelectList(Type enumType)` if people are passionate about that. 

one requirement for placing that method is that it have access to an `IModelMetadataProvider`.  probably easiest to make it an `IHtmlHelper` instance method.  unfortunately a `static` method in `SelectList` won't meet that requirement (passing in the provider is ugly since user code rarely if ever needs it).

**Comment by dougbu** [comment]
- fixed in commits 7dd3afe and 570b1e5
- updated subject to match final change
